Code Quality: Improve prefix, better scope, for animations

jasmussen opened the pull request:
This PR maybe fixes #11668. At least, this is the intention. It does two things:

- It changes the naming scheme for all animations. Instead of having a mix of underscores and dashes for word separation, it uses BEM (to the best of my ability) to add consistency and a new naming convention for all animations.
- It adds a prefix to all animations, indicating they are for the editor. 
- It moves all keyframe animations out of the mixins file, and into the edit-post style. This is because keyframe definitions don't work well with mixins and are just duplicated.

This PR has been tested for all the animations I could find that were using the ones defined. But please give me a sanity check.

By the way I noticed the following two animations do not appear to be used:

- `editor-animation__animate-fade`
- `editor-animation__move-background`

Should we remove those? Or can anyone recall where they were intended to be used?

Another question, right now these animations are prefixed `editor-animation`. Should we find a prefix that indicates these might be used outside of the editor, in the rest of the admin, one day?

Finally — what is the risk associated with changing these animation names? Do we have any way to know if plugin developers might have used these?

Commit:
Code Quality: Improve prefix, better scope, for animations

This PR maybe fixes #11673, maybe fixes #11668. At least, this is the intention. It does two things:

- It changes the naming scheme for all animations. Instead of having a mix of underscores and dashes for word separation, it uses BEM (to the best of my ability) to add consistency and a new naming convention for all animations.
- It adds a prefix to all animations, indicating they are for the editor. Though perhaps we should have a more _admin_ specific prefix given these might one day be used outside the admin?
- It moves all keyframe animations out of the mixins file, and into the edit-post style. This is because keyframe definitions don't work well with mixins and are just duplicated.

This PR has been tested for all the animations I could find that were using the ones defined. But please give me a sanity check.

By the way I noticed the following two animations do not appear to be used:

- editor-animation__animate-fade
- editor-animation__move-background

Should we remove those? Or can anyone recall where they were intended to be used?

youknowriad commented:
I'd personally suggest removing this file for the auto-imported files, it doesn't make sense to include in a lot of generated stylesheets. Let's make it an explicit import.

jasmussen commented:
Edited PR, I don't think this fix issue #11673.

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,42 @@
-
-@mixin animate_fade {
-	animation: animate_fade 0.1s ease-out;
+@mixin editor-animation__animate-fade {
```
This doesn't seem to be used, I'll just remove

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,42 @@
 }
 
-@mixin move_background {
+@mixin editor-animation__move-background {
```
This doesn't seem to be used either, I'd consider removing too.

jasmussen replied:
There's this test for this specific animation:

```
	it( 'should ignore keyframes', () => {
		const callback = wrap( '.my-namespace' );
		const input = `
		@keyframes editor-animation__move-background {
			from {
				background-position: 0 0;
			}
		}`;
		const output = traverse( input, callback );

		expect( output ).toMatchSnapshot();
	} );
```

I agree it doesn't seem to be used, but I'm also not quite sure either what this test does or what to do with it. Any suggestions?

youknowriad replied:
you can just replace it with any animation (defined or not)

youknowriad replied:
To clarify, it's only testing that this editor styles wrapping behavior doesn't touch the animations.

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,42 @@
 }
 
-@mixin slide_in_right {
+@mixin editor-animation__slide-in-right {
```
This is used only once, I'd consider moving it to the place it's used :)

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,42 @@
 }
 
-@mixin slide_in_top {
+@mixin editor-animation__slide-in-top {
```
Also used once.

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,42 @@
-
-@mixin region_focus($speed: 0.2s) {
-	animation: editor_region_focus $speed ease-out;
+@mixin editor-animation__region-focus($speed: 0.2s) {
```
Also used once, and seem very specific to region focus, so just move it there.

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,42 @@
-	to {
-		margin-top: 0;
-	}
+@mixin editor-animation__rotation($speed: 1s) {
```
Also used once.

jdevalk commented:
Looking very good, also agree with all of @youknowriad's comments, removing a lot of these would make everything smaller :)

Mostly happy to see that if you work through his comments I think we're only left with 3 `@mixin`s. Not 100% sure how this works in Gutenberg but would this remove them from `block-library/style.css` entirely?

jasmussen commented:
Good feedback, will address all points in a bit. 

> Not 100% sure how this works in Gutenberg but would this remove them from block-library/style.css entirely?

I tested reloading that specific CSS file in a separate tab to verify this worked, and yes indeed moving the keyframe rules from the mixins file removes them from the block-library/style.css file.

One note for perhaps a separate ticket — `/* These comments */` are not stripped from minified CSS files, whereas `// These comments` _are_ stripped. Which means there are a bunch of empty heading-specific CSS comments there, even though perhaps they shouldn't be.

Commit:
Address feedback.

This retires a few animations, and moves others that are only used once to their respective scopes.

Commit:
Address comment.

Gallery related, in https://github.com/WordPress/gutenberg/issues/11668#issuecomment-438007004.

jasmussen commented:
Okay, I think I addressed everything.

jdevalk commented:
👍 except for the failing build ;)

jasmussen commented:
Restarting, hoping it passes this time :D There's been some travis shenanigans lately.

youknowriad commented on packages/components/src/spinner/style.scss:
```diff
@@ -19,6 +19,15 @@
 		border-radius: 100%;
 		transform-origin: 6px 6px;
-		@include animate_rotation;
+		animation: editor-animation__rotation 1s infinite linear;
```
Could we give it a name using our guidelines: `components-spinner__animation` ?

youknowriad commented on packages/edit-post/src/components/fullscreen-mode/style.scss:
```diff
@@ -19,9 +19,16 @@ body.is-fullscreen-mode {
+	}
+}
+
+@keyframes editor-animation__slide-in-top {
```
could this be `edit-post-fullscreen-mode__slide-in-animation`?

youknowriad commented on packages/edit-post/src/components/layout/style.scss:
```diff
@@ -128,14 +128,21 @@
 	}
 }
 
+@keyframes editor-animation__slide-in-right {
```
This could be `edit-post-layout__slide-in-animation`

youknowriad commented on packages/edit-post/src/style.scss:
```diff
@@ -20,28 +20,15 @@
+// These keyframes should not be part of the _animations.scss mixins file.
+// Because keyframe animations can't be defined as mixins properly, they are duplicated.
+// Since hey are intended only for the editor, we add them here instead.
+@keyframes editor-animation__loading-fade {
```
maybe more `edit-post__loading-fade-animation`

youknowriad commented on assets/stylesheets/_animations.scss:
```diff
@@ -1,78 +1,13 @@
-
-@mixin modal_appear() {
-	animation: modal-appear 0.1s ease-out;
+@mixin editor-animation__modal-appear() {
```
This mixin doesn't seem useful as use only once, we can move it and rename it as well?

Commit:
Address feedback.

jasmussen commented:
Okay I think I got all those fixed. Final look?

youknowriad reviewed (approved):
LGTM 👍 Thanks for the updates @jasmussen Feel free to merge after the last minor comment.

Commit:
Address final point.

jasmussen commented on packages/components/src/modal/style.scss:
```diff
@@ -40,7 +40,17 @@
+	}
+}
+
+@keyframes edit-post__modal-appear-animation {
```
Haha, it'll stick on my brain any day now! Thank you.
